### ISSUE TITLE
[Operator composition, adjoint, out] somehow not working

### ISSUE THREAD

**Issue by MatthiasJE** [odlgroup/odl#1060]
Based on the setting

```
import odl

domain = odl.uniform_discr([0, 0], [1, 1], [100, 100])
D = odl.Gradient(domain)
s = odl.operator.MultiplyOperator(D.domain.one(), domain=D.range, range=D.range)

M = s * D

x = M.domain.one()
```
the call `M.adjoint(M.range.one())` works and `M.adjoint(M.range.one(), out=x)` does not. It gives an error message saying that elements are not in the appropriate spaces. Does someone have a clue what is going wrong? To me everything is well defined here.

As we are having this example here already, why is `D.domain.one() * D` not well defined in odl? It should do the same as M above.

**Comment by kohr-h** [comment]
The problem here is that broadcasting doesn't happen automatically. Formally the operator `S` has an incompatible multiplicand since it multiplies a given element of `D.range`, which is a product space, with an element of `domain`.

Now, weird enough, two different things happen depending on whether or not `out is None`.

### `out is None`
This case ends up calling [this line](https://github.com/odlgroup/odl/blob/6f60c9d0466634677510bc94c41e66d885d4c764/odl/operator/default_ops.py#L269), and by some magic happening [here](https://github.com/odlgroup/odl/blob/9b401dbb2b1bf41b5de783f0ef1ed433b6a82487/odl/space/pspace.py#L936-L986) these methods (meaning `__mul__` for this case) all support broadcasting. So stuff works as expected.

### `out is not None`
In this case, we get to [this line](https://github.com/odlgroup/odl/blob/6f60c9d0466634677510bc94c41e66d885d4c764/odl/operator/default_ops.py#L274) which uses the default `LinearSpace` implementation. That one doesn't know anything specific and thus requires the operands to be in the same space. [Crash!](https://github.com/odlgroup/odl/blob/bab273aa32d163603514b8646b3cc6fd58b82ce7/odl/set/space.py#L324-L326)

 ## How to fix it
Either `ProductSpace` must override `multiply` to allow broadcasting, or `MultiplyOperator` must check if there is need for broadcasting and act accordingly. Opinions, @adler-j?

**Comment by MatthiasJE** [comment]
The short term solution that I will use is
```
s1 = odl.operator.MultiplyOperator(x, domain=D.domain, range=D.domain)
s2 = odl.DiagonalOperator(s1, s1)
M = s2 * D
```
but this is obviously not very beautiful. Why can't this be implemented as `M = x * D`?

**Comment by kohr-h** [comment]
I agree that this would be beautiful, but while it looks easy and very logical, it's not at all trivial to get working in a consistent and reliable way.

The problem with the expression `M = x * D` is that formally they're incompatible. Currently the expression gets transformed to (roughly) `M = MultiplyOperator(x) * D` as above, so far so good. But now `D.range` does not match the domain of the multiply op, so we get an error.

To make this work, one of the operands has to be aware of the need for broadcasting. The left operand has no chance to do that because it's not a product space element (or op defined on a pspace later) so it cannot do anything.
The right guy is the one that can do something because its domain is a pspace. Currently it behaves as any other operator, and modifying that would be a fundamental change. (There are quite a number of other operators that would need something similar, basically all that are associated with binary operations.)
But assuming we want to fix `MultiplyOperator` in isolation, we need to add a case in the case `out is not None` that uses multiplication with broadcasting instead of `x.multiply` if necessary.
Alternatively, we can make `x.multiply` broadcasting-aware.

In any case, this stuff needs some more thought, see also #1062.

**Comment by kohr-h** [comment]
Is there more to do here? I'm not sure if we want a fundamental change to operators just for a slightly nicer syntax.
